js/core/editorJavaScript.js

```diff
@@ -47,6 +47,30 @@
         }
       }
     });
+    Espruino.Core.Config.add("INDENTATION_TYPE", {
+      section : "General",
+      name : "Indentation Type",
+      description : "Whether to indent using spaces or tab characters",
+      type : { "spaces": "Spaces", "tabs": "Tabs" },
+      defaultValue : "spaces",
+      onChange : function(newValue) {
+        if (codeMirror) {
+          codeMirror.setOption('indentWithTabs', !!(Espruino.Config.INDENTATION_TYPE == "tabs"));
+        }
+      }
+    });
+    Espruino.Core.Config.add("TAB_SIZE", {
+      section : "General",
+      name : "Indentation Size",
+      description : "The number of space characters an indentation should take up",
+      type : {1:1,2:2,4:4,8:8},
+      defaultValue : 2,
+      onChange : function(newValue) {
+        if (codeMirror) {
+          codeMirror.setOption('indentUnit', Espruino.Config.TAB_SIZE);
```

Comment:
Should this handler also set `tabSize` since that's set on line 106 as well?

Reply:
Having read the code you highlighted, that could well be the issue... Strange how CodeMirror implements two separate properties when changing tab size! Will take a look tonight and see if that's the cause of the crazy indentation problems.